url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/NestAgent.java

```diff
@@ -0,0 +1,363 @@
+
+	// MESSAGES
+	@Override
+	public void msgHereIsPartType(PartType type) {
+		print("Received msgHereIsPartType");
+		if (currentPartType == null && type == null) {
+			currentPartType = type;
+		}
+		else if(currentPartType == null && type != null) {
+			//if(!type.equals(currentPartType)) {
+				currentPartType = type;
+				state = NestState.PURGING;
+				laneState = LaneState.PURGING;
+				lane.msgPurgeParts();
+			//} 
+		}
+		else if(currentPartType != null && type != null) {
+			if(!type.equals(currentPartType)) {
+				currentPartType = type;
+				state = NestState.PURGING;
+				laneState = LaneState.PURGING;
+				lane.msgPurgeParts();
+			} else {
+				state = NestState.DONE_PURGING;
+				takingParts=false;
+			}
+		} else if (currentPartType != null) {
+			currentPartType = type;
+			state = NestState.PURGING;
+			laneState = LaneState.PURGING;
+			lane.msgPurgeParts();
+		}
+
+		// camera.msgResetSelf();
+
+		stateChanged();
+	}
+
+	@Override
+	public void msgPartReady() {
+		print("Received msgPartReady from lane");
+		partReady = true;
+		stateChanged();
+
+	}
+
+	@Override
+	public void msgHereIsPart(Part p) {
+		print("Received msgHereIsPart");
+		takingParts = false;
+		countRequest--;
+		count++;
+		// if (currentParts.size() <= full) {
+		currentParts.add(new MyPart(p));
+		print("Received a part of type " + p.type.getName() + " I have " + currentParts.size()
+				+ " parts and have requested " + countRequest + " takingParts: " + takingParts);
+		// }
+		stateChanged();
+	}
+
+	@Override
+	public void msgTakingPart(Part p) {
+		print("Received msgTakingPart");
+		synchronized (currentParts) {
+			for (MyPart part : currentParts) {
+				if (part.part.equals(p)) {
+					print("found part");
+					part.status = NestStatus.REMOVING;
+					// currentParts.remove(part);
+					break;
+				}
+			}
+		}
+		stateChanged();
+	}
+
+	@Override
+	public void msgLanePurgeDone() {
+		print("Received msgLanePurgeDone");
+		state = NestState.DONE_PURGING;
+		if (currentPartType == null) {
+			state = NestState.NULL;
+		}
+		laneState = LaneState.READY;
+		stateChanged();
+	}
+
+	@Override
+	public void msgPurgeSelf() {
+		state = NestState.PRIORITY_PURGE;
+		stateChanged();
+	}
+
+	@Override
+	public void msgDoneTakingParts() {
+		print("Received msgDoneTakingParts");
+		count--;
+		takingParts = false;
+		stateChanged();
+	}
+
+	@Override
+	public void msgReceivePartDone() {
+		print("Received msgReceivePartDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgGivePartToPartsRobotDone() {
+		print("Received msgGivePartToPartsRobotDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgPurgingDone() {
+		print("Received msgPurgingDone from graphics");
+		state = NestState.DONE_PURGING;	
+		if(currentPartType == null) {
+			state = NestState.NULL;
+		}
+		animation.release();
+		stateChanged();
+	}
+	
+	@Override
+	public boolean pickAndExecuteAnAction() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.